osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -120,44 +118,111 @@ private DbEncryptedSecret createEncrypted(final String deviceIdentification, fin
+        final EncryptedSecret encryptedSecret = new EncryptedSecret(keyReference.getEncryptionProviderType(),
+                secretBytes);
+        return this.createTypedSecret(dbEncryptedSecret, keyReference, encryptedSecret);
     }
 
     private TypedSecret createTypedSecret(final DbEncryptedSecret dbEncryptedSecret,
-                                          final DbEncryptionKeyReference keyReference, final EncryptedSecret encryptedSecret) {
+            final DbEncryptionKeyReference keyReference, final EncryptedSecret encryptedSecret) {
         try {
-            final Secret decryptedSecret = this.encryptionDelegate.decrypt(encryptedSecret,
+            final byte[] decryptedSecret = this.encryptionDelegate.decrypt(encryptedSecret,
                     keyReference.getReference());
             final TypedSecret typedSecret = new TypedSecret();
-            typedSecret.setSecret(HexUtils.toHexString(decryptedSecret.getSecret()));
+            typedSecret.setSecret(HexUtils.toHexString(decryptedSecret));
             typedSecret.setSecretType(dbEncryptedSecret.getSecretType());
             return typedSecret;
         } catch (final Exception exc) {
             throw new IllegalStateException("Could not decrypt secret (id: " + dbEncryptedSecret.getId() + ")", exc);
         }
     }
+
+    public void storeSecrets(final String deviceIdentification, final List<TypedSecret> secrets) {
+        //@formatter:off
+        secrets.stream()
+                .map(t -> this.validateNewSecret(deviceIdentification, t))
+                .map(s -> this.createEncrypted(deviceIdentification, s, this.getKey()))
+                .forEach(this.secretRepository::save);
+        //@formatter:on
+    }
+
+    public synchronized void activateNewSecrets(final String deviceIdentification, final List<SecretType> secretTypes) {
+        secretTypes.stream().forEach(t -> this.checkNrNewSecretsOfType(deviceIdentification, t, 1));
+        secretTypes.stream().forEach(t -> this.activateNewSecret(deviceIdentification, t));
+    }
+
+    public boolean hasNewSecret(final String deviceIdentification, final SecretType secretType) {
+        return this.secretRepository.getSecretCount(deviceIdentification, secretType, SecretStatus.NEW) > 0;
+    }
+
+    private synchronized long activateNewSecret(final String deviceIdentification, final SecretType secretType) {
+        final Optional<DbEncryptedSecret> activeSecretOptional = this.getSingleDbEncryptedSecret(deviceIdentification,
+                secretType, SecretStatus.ACTIVE);
+        if (activeSecretOptional.isPresent()) {
+            final DbEncryptedSecret currentSecret = activeSecretOptional.get();
+            currentSecret.setSecretStatus(SecretStatus.EXPIRED);
+            this.secretRepository.save(currentSecret);
+        }
+        final DbEncryptedSecret newSecret = this.getSingleDbEncryptedSecret(deviceIdentification, secretType,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Call "Optional#isPresent()" before accessing the value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS3655)